Implement the sharing popup UI.

kadams54 opened the pull request:
The UI isn't hooked up to data yet; that'll be the next chunk of work.

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6668/

kadams54 commented:
## QA

1. Modify `jujugui/views.py` and set the `shareFlag` to true.
2. `make run` and open the GUI at http://0.0.0.0:6543.
3. You should see a share button in the canvas header; click on it.
4. You should see a sharing popup appear with some dummy users (complete with broken icons) listed. Compare against the design: ![Sharing Design](https://cloud.githubusercontent.com/assets/2169216/13116832/593f9b24-d595-11e5-962e-a97bac4e58dc.png)

Repeat the same testing of the popup for the sharing icons in the model list at `/u/admin`.

kadams54 commented on jujugui/static/gui/src/app/components/sharing/sharing.js:
```diff
@@ -0,0 +1,141 @@
+    _getModelUserInfoCallback: function(error, data) {
+      this.setState({loadingUsers: false}, () => {
+        if (error) {
+          // TODO kadams54: figure out a better way to handle this.
```
These TODOs will be implemented when the real data is wired in.

kadams54 commented on jujugui/static/gui/src/app/components/sharing/sharing.js:
```diff
@@ -0,0 +1,141 @@
+      @returns {Array} An array of markup objects for each user.
+    */
+    _generateUsersWithAccess: function() {
+      // TODO kadams54: display a spinner instead when loading data.
```
Will be implemented when the real data is wired in.

frankban commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -2634,6 +2636,36 @@ YUI.add('juju-gui', function(Y) {
     },
 
+    // TODO: Move to views/utils.js.
+    showSharing: function(show = true) {
```
Please add a docstring, and why not just placing it in utils now?

hatched replied:
It's a little weird that a method called `showSharing` can take an argument to do the exact opposite of the name :)

What about `sharingVisibility`?

kadams54 replied:
Yeah, I was supposed to make that change before making the PR. Will fix and add a docstring.

frankban commented on jujugui/static/gui/src/app/app.js:
```diff
@@ -2634,6 +2636,36 @@ YUI.add('juju-gui', function(Y) {
+      const modelUserInfo = function(callback) {
+        callback(false, [
+          {
+            icon: 'who.png',
```
AFAIK we don't have icons, as the ModelUserInfo response is a list of 
```go
type ModelUserInfo struct {
	UserName       string               `json:"user"`
	DisplayName    string               `json:"display-name"`
	LastConnection *time.Time           `json:"last-connection"`
	Access         UserAccessPermission `json:"access"`
}
```
I guess access is "read", "write" or "admin". Also, each result can have an error.
I'd mock the future call like 
```javascript
callback(null, [{
  name: 'who',
  displayName: 'Dr. Who',
  lastConnection: 'now',
  access: 'admin'
}, {
  name: 'dalek',
  displayName: 'Dalek',
  lastConnection: 'never',
  access: 'write',
  err: 'exterminate!'
}]);
```

hatched commented on jujugui/static/gui/src/app/components/sharing/sharing.js:
```diff
@@ -0,0 +1,141 @@
+  juju.components.Sharing = React.createClass({
+    propTypes: {
+      closeHandler: React.PropTypes.func,
+      modelUserInfo: React.PropTypes.func.isRequired
```
`getModelUserInfo` ?

hatched commented on jujugui/static/gui/src/app/components/sharing/sharing.js:
```diff
@@ -0,0 +1,141 @@
+        return (
+          <div key={user.username} className="sharing__user">
+            <div className="sharing__user-icon">
+              <img src={user.icon}/>
```
It's unlikely this functionality will be implemented any time soon so you might as well remove it.

hatched commented on jujugui/static/gui/src/app/components/sharing/sharing.js:
```diff
@@ -0,0 +1,141 @@
+              {user.username}
+            </div>
+            <div className="sharing__user-name">
+              {user.name}
```
I dont' think we have this information either.

kadams54 replied:
I'll adjust as necessary once I get the real data wired in and know what I have and don't have. Same goes for the icon.

frankban commented on jujugui/static/gui/src/app/components/sharing/sharing.js:
```diff
@@ -0,0 +1,141 @@
+
+YUI.add('sharing', function() {
+
+  juju.components.Sharing = React.createClass({
```
Please add a dcostring describing the behavior and the goals of this component.

frankban commented on jujugui/static/gui/src/app/components/sharing/sharing.js:
```diff
@@ -0,0 +1,141 @@
+    */
+    _getModelUserInfo: function() {
+      this.setState({loadingUsers: true}, () => {
+        const xhr = this.props.modelUserInfo(this._getModelUserInfoCallback);
```
This is not an xhr, this is a WebSocket call, and I don't think that modelUserInfo will return anything. What am I missing?

kadams54 replied:
I suspect most of these comments will be worked on in the next PR, once the real data is hooked up.

frankban commented on jujugui/static/gui/src/app/components/sharing/sharing.js:
```diff
@@ -0,0 +1,141 @@
+
+      @method _getModelUserCallback
+    */
+    _getModelUserInfoCallback: function(error, data) {
```
s/data/users/ ?

frankban commented on jujugui/static/gui/src/app/components/sharing/sharing.js:
```diff
@@ -0,0 +1,141 @@
+          console.error('cannot fetch model user info', error);
+          return;
+        }
+        // TODO kadams54: transform into the expected user objects.
```
What's the expected user object? We need to extend each user dict? Or change the name for the keys?

kadams54 replied:
Don't know. It's mostly just a placeholder in case any data munging needs to happen between getting the data from the API and then passing it along for display in the GUI. Another place that will be hashed out once the real data is in place.

frankban commented on jujugui/static/gui/src/app/components/sharing/sharing.js:
```diff
@@ -0,0 +1,141 @@
+        return;
+      }
+      return users.map((user) => {
+        const roleMarkup = user.role ? (
```
s/role/access/ ?

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6670/

bac commented:
I followed the QA instructions.

Using Safari, clicking on the share button didn't do anything. No pop up. Got this in the error console:
```
Invariant Violation: _registerComponent(...): Target container is not a DOM element.
```

When clicking on the share button in `/u/admin` I do get the popup.

On Chrome, the popup appears in both scenarios.

In the QA instructions it says broken icons will be shown. They did not appear. In Safari and Chrome I see the following, which doesn't match the design closely:

<img width="769" alt="screen shot 2017-02-06 at 7 30 36" src="https://cloud.githubusercontent.com/assets/617845/22647342/eec86976-ec3e-11e6-9df5-23c87690454e.png">

kadams54 commented:
@bac: I tested in Safari and was not able to replicate the error. Maybe the old HTML (which would not have the 'sharing-container' element and would throw that error) was being cached?

Apologies on the (unspecified) deviations from design. I've updated the QA instructions to be more current. Specifically, icons were dropped due to suggestions from @frankban and @hatched. Additionally, I've found out that we don't have backend support for sharing from the profile, so the button is removed from there and is now only in the header. QA instructions are also updated to reflect these changes.

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6683/

bac commented:
QA looks good this time, except for the missing Cancel button, which I suspect is by design...or deviation from design.

kadams54 commented:
@bac: Thanks! Yes, you are correct. The "Cancel" button isn't there right now since this is a read-only modal. Functionally it would do exactly the same thing as the "Done" button. Once the ability to invite new people is added, the "Cancel" button will be re-enabled.

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6689/

jujugui commented:
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6690/

kadams54 commented:
Thanks all. :shipit:

jujugui commented:
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

spencerbygraves commented:
As we are only listing users, we'd like to make a few small updates to the design. We've discussed these changes with @Lukewh 

- Change the title to 'Shared with'
- Change 'Users with access' to 2 column headers - User and Access
- Use our base button style with the label 'Close'
- Make the button full width
- Scrollbar aligned to the edge as on the inspector
- Include a bit more white space between columns
